src/llvm/mod.rs

```diff
@@ -179,69 +181,130 @@ pub unsafe fn optimize(
+    // calls to the equivalent functions, but we can't remove the instructions
+    // until we're done iterating over them.
+    let mut mem_intrinsic_instructions = Vec::new();
+    for function in module.functions_iter() {
+        for basic_block in function.basic_blocks_iter() {
+            for instruction in basic_block.instructions_iter() {
+                let instruction = LLVMIsAMemIntrinsic(instruction);
+                if instruction.is_null() {
+                    continue;
+                }
+                mem_intrinsic_instructions.push(instruction);
+            }
+        }
+    }
+
+    // LLVMGet* functions do not pass ownership to the caller.
+    let context = LLVMGetModuleContext(module);
+
+    let builder = LLVMCreateBuilderInContext(context);
+    for instruction in mem_intrinsic_instructions.iter().copied() {
+        let instruction_num_operands = LLVMGetNumOperands(instruction);
+        let instruction_num_operands: u32 = instruction_num_operands.try_into().unwrap();
+        let mut instruction_operands: Vec<_> = (0..instruction_num_operands)
+            .map(|i| LLVMGetOperand(instruction, i))
+            .collect();
+
+        let instruction_name = if !LLVMIsAMemCpyInst(instruction).is_null() {
+            "memcpy"
+        } else if !LLVMIsAMemMoveInst(instruction).is_null() {
+            "memmove"
+        } else if !LLVMIsAMemSetInst(instruction).is_null() {
+            "memset"
+        } else {
+            panic!("unknown mem intrinsic");
+        };
+        let function = {
+            let instruction_name = CString::new(instruction_name).unwrap();
+            LLVMGetNamedFunction(module, instruction_name.as_ptr())
+        };
+        // The user neglected to defined a required intrinsic. Perhaps we
+        // should return an error instead of panicking.
+        assert!(
+            !function.is_null(),
+            "missing mem intrinsic function {instruction_name}"
+        );
+        let function_type = LLVMGlobalGetValueType(function);
+        let function_type = {
+            let return_type = LLVMGetReturnType(function_type);
+            let param_count = LLVMCountParamTypes(function_type);
+            // Use the parameter types we're about to pass rather than the parameter types of the function.
+            let mut param_types: Vec<_> = instruction_operands
+                .iter()
+                .copied()
+                .map(|operand| LLVMTypeOf(operand))
+                .collect();
+            let is_var_arg = LLVMIsFunctionVarArg(function_type);
+
+            LLVMFunctionType(
+                return_type,
+                param_types.as_mut_ptr(),
+                param_count,
+                is_var_arg,
+            )
+        };
+
+        LLVMPositionBuilderBefore(builder, instruction);
+        // Call instructions can't have a name; attempting to set one crashes LLVM.
+        let name = [0];
```

Comment:
nit: it means the same thing but might be more obvious if you wrote this `['\0']`

Reply:
This should be more obvious now.